### PR TITLE
Clarify error message when this is returned by reference

### LINKED ISSUE #15704: Misleading error message in case of ref-returning this in a reference type
**Version Used**:  https://github.com/dotnet/roslyn/commit/0694d5ccc05b7e1cc37c82a1c15c8f9fc23e3775

**Steps to Reproduce**:

```cs
public class C 
{
    public ref C M()
    {
        return ref this;
    }
}
```
**Expected Behavior**:

`error CS1605: Cannot use 'this' as a ref or out value because it is read-only`

**Actual Behavior**:

`error CS8170: Struct members cannot return 'this' or other instance members by reference`

### PR DESCRIPTION
**Pull request by alrz**
Fixes #15704

### CODE REVIEW

**Comment by alrz** [comment]
I wonder if I can mention teams like this @dotnet/roslyn-compiler :D

Please review.

**Comment by alrz** [comment]
Can I have a review on this?

**Comment by gafter** [comment]
I don't see any changes in the diagnostics. It looks like #15704 has already been fixed. Based on that, I would reject this PR. Am I missing something?

**Comment by alrz** [comment]
@gafter I could repro #15704 on https://github.com/dotnet/roslyn/commit/874c89a8a7712f768c5874a0296969e908e59ab7 (yesterday)

[tryroslyn](http://tryroslyn.azurewebsites.net/#b:master/MYGwhgzhAEDCBQBve1XQE4FMBmdoFkAKASmgF4A+DHaAFwAsBLCAblXgF8g=)

**Comment by gafter** [comment]
I'm just looking at the diffs in the error messages in your tests. No differences, just the additional error from the additional code you added in the test.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturnsTests.cs [hunk 1]
```diff
@@ -595,37 +595,46 @@ public struct S1
...
+                Diagnostic(ErrorCode.ERR_RefReturnCall, "Foo(ref M2.x)").WithArguments("Test.Foo<char>(ref char)", "arg").WithLocation(41, 24),
+                // (46,32): error CS8168: Cannot return local 'M2' by reference because it is not a ref local
+                //             return ref Foo(ref M2).x;
+                Diagnostic(ErrorCode.ERR_RefReturnLocal, "M2").WithArguments("M2").WithLocation(46, 32),
+                // (46,24): error CS8165: Cannot return by reference a member of result of 'Test.Foo<Test.S1>(ref Test.S1)' because the argument passed to parameter 'arg' cannot be returned by reference
+                //             return ref Foo(ref M2).x;
+                Diagnostic(ErrorCode.ERR_RefReturnCall2, "Foo(ref M2)").WithArguments("Test.Foo<Test.S1>(ref Test.S1)", "arg").WithLocation(46, 24),
+                // (58,24): error CS1605: Cannot use 'this' as a ref or out value because it is read-only
+                //             return ref this;
+                Diagnostic(ErrorCode.ERR_RefReadonlyLocal, "this").WithArguments("this").WithLocation(58, 24));
```
Comment (alrz) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturnsTests.cs]: @gafter This case did not have a test, I added it to `RefReadonlyCall`. Should I separate it out?

Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturnsTests.cs]: Its fine... are you saying the diagnostics for this case were different before?

Comment (alrz) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturnsTests.cs]: Yes.

**Comment by VSadov** [comment]
Should be rebased onto post-dev15
Possibly can just change the base branch of the PR in the GitHub.

**Comment by VSadov** [comment]
@dotnet-bot test this please

Changed the base branch.

**Comment by gafter** [comment]
(closed and reopened to trigger retest)

**Comment by gafter** [comment]
@alrz Can you please merge with master to bring this PR up-to-date?

**Comment by alrz** [comment]
@gafter will do. out of curiosity, as long as there is conflict, would that make any difference?

**Comment by gafter** [comment]
@alrz No idea. But now the tests pass.

**Comment by MeiChin-Tsai** [comment]
approved. thx.

**Comment by alrz** [comment]
@gafter Is there a chance to have this merged to dev15.6? rebase?

**Comment by alrz** [comment]
I rebased to 15.6 just in case you want this for the next update.

**Comment by sharwell** [comment]
@dotnet-bot retest this please

**Comment by sharwell** [comment]
@MeiChin-Tsai @jaredpar Since there was a delay between last ask and right now, can you confirm the approval? I'm rerunning all tests now against master as a sanity check too.

**Comment by jaredpar** [comment]
Yes, still approved.  Would not make escrow bar though.

**Comment by sharwell** [comment]
@jaredpar Thank you. I restarted the timeout builds and if I see a clean one today I'll merge it.

**Comment by MeiChin-Tsai** [comment]
Please try to get it in asap. Still approved.

**Pull request merged**